File: src/codex_broker/codex/adapter.py
from dataclasses import dataclass
from typing import Any

from ..domain.models import LoginMethod
from .client import AppServerClient


@dataclass(frozen=True, slots=True)
class Secret:
    _value: str

    def reveal(self) -> str:
        return self._value

    def __repr__(self) -> str:
        return "Secret('[REDACTED]')"

    def __str__(self) -> str:
        return "[REDACTED]"


@dataclass(frozen=True, slots=True)
class LoginInteraction:
    login_id: str
    method: LoginMethod
    auth_url: Secret | None = None
    verification_url: Secret | None = None
    user_code: Secret | None = None
    expires_at_ms: int | None = None


class CodexAdapter:
    def __init__(self, client: AppServerClient) -> None:
        self.client = client

    async def start_login(self, method: LoginMethod) -> LoginInteraction:
        if method == LoginMethod.MANUAL_TOKENS:
            raise ValueError("manual token import does not start OAuth")
        kind = "chatgpt" if method == LoginMethod.CHATGPT_BROWSER else "chatgptDeviceCode"
        params: dict[str, Any] = {"type": kind}
        if method == LoginMethod.CHATGPT_BROWSER:
            params |= {"useHostedLoginSuccessPage": True, "appBrand": "codex"}
        result, _ = await self.client.request("account/login/start", params)
        return LoginInteraction(
            login_id=str(result.get("loginId", "")),
            method=method,
            auth_url=Secret(result["authUrl"]) if result.get("authUrl") else None,
            verification_url=Secret(result["verificationUrl"])
            if result.get("verificationUrl")
            else None,
            user_code=Secret(result["userCode"]) if result.get("userCode") else None,
            expires_at_ms=result.get("expiresAt"),
        )

    async def cancel_login(self, login_id: str) -> None:
        await self.client.request("account/login/cancel", {"loginId": login_id})

    async def account(self, refresh_token: bool = False) -> dict[str, Any]:
        result, _ = await self.client.request("account/read", {"refreshToken": refresh_token})
        return result

    async def rate_limits(self) -> dict[str, Any]:
        result, _ = await self.client.request("account/rateLimits/read", {})
        return result

    async def pulse_windows(self) -> dict[str, Any]:
        listed, _ = await self.client.request("model/list", {"limit": 100})
        model, effort = self._pulse_model(listed.get("data"))
        started, _ = await self.client.request(
            "thread/start",
            {"ephemeral": True, "model": model, "serviceTier": "default"},
        )
        thread_id = started.get("thread", {}).get("id")
        if not isinstance(thread_id, str) or not thread_id:
            raise RuntimeError("Codex returned no window-pulse thread ID")
        turn, _ = await self.client.request(
            "turn/start",
            {
                "threadId": thread_id,
                "input": [{"type": "text", "text": "Reply OK."}],
                "model": model,
                "effort": effort,
                "serviceTier": "default",
            },
        )
        await self._wait_for_turn(str(turn.get("turn", {}).get("id", "")))
        return await self.rate_limits()

    @staticmethod
    def _pulse_model(data: Any) -> tuple[str, str]:
        candidates: list[tuple[int, str, str]] = []
        for item in data if isinstance(data, list) else []:
            if not isinstance(item, dict) or item.get("hidden"):
                continue
            model = item.get("model")
            modalities = item.get("inputModalities", [])
            efforts = item.get("supportedReasoningEfforts", [])
            names = [
                value.get("reasoningEffort")
                for value in efforts
                if isinstance(value, dict)
            ]
            if isinstance(model, str) and "text" in modalities and names:
                effort = "minimal" if "minimal" in names else "low" if "low" in names else names[0]
                if isinstance(effort, str):
                    candidates.append((0 if "mini" in model else 1, model, effort))
        if not candidates:
            raise RuntimeError("Codex returned no text model for window pulse")
        _, model, effort = min(candidates)
        return model, effort

    async def _wait_for_turn(self, turn_id: str) -> None:
        if not turn_id:
            raise RuntimeError("Codex returned no window-pulse turn ID")
        async for message in self.client.notifications():
            if message.get("method") != "turn/completed":
                continue
            turn = (message.get("params") or {}).get("turn")
            if not isinstance(turn, dict) or turn.get("id") != turn_id:
                continue
            if turn.get("status") != "completed":
                raise RuntimeError("Codex window pulse failed")
            return
        raise RuntimeError("Codex closed before window pulse completed")

File: internal/codex/adapter.go
package codex

import (
	"context"
	"errors"
	"sort"
)

type LoginInteraction struct {
	LoginID, Method, AuthURL, VerificationURL, UserCode string
	ExpiresAtMS                                         int64
}

type Adapter struct{ Client *Client }

func (a Adapter) StartLogin(ctx context.Context, method string) (LoginInteraction, error) {
	if method == "MANUAL_TOKENS" {
		return LoginInteraction{}, errors.New("manual token import does not start OAuth")
	}
	kind := "chatgptDeviceCode"
	params := map[string]any{"type": kind}
	if method == "CHATGPT_BROWSER" {
		params = map[string]any{"type": "chatgpt", "useHostedLoginSuccessPage": true, "appBrand": "codex"}
	}
	result, err := a.Client.Request(ctx, "account/login/start", params)
	if err != nil {
		return LoginInteraction{}, err
	}
	return LoginInteraction{LoginID: text(result["loginId"]), Method: method, AuthURL: text(result["authUrl"]), VerificationURL: text(result["verificationUrl"]), UserCode: text(result["userCode"]), ExpiresAtMS: integer(result["expiresAt"])}, nil
}

func (a Adapter) CancelLogin(ctx context.Context, loginID string) error {
	_, err := a.Client.Request(ctx, "account/login/cancel", map[string]any{"loginId": loginID})
	return err
}

func (a Adapter) Account(ctx context.Context, refresh bool) (map[string]any, error) {
	return a.Client.Request(ctx, "account/read", map[string]any{"refreshToken": refresh})
}

func (a Adapter) RateLimits(ctx context.Context) (map[string]any, error) {
	return a.Client.Request(ctx, "account/rateLimits/read", map[string]any{})
}

func (a Adapter) PulseWindows(ctx context.Context) (map[string]any, error) {
	listed, err := a.Client.Request(ctx, "model/list", map[string]any{"limit": 100})
	if err != nil {
		return nil, err
	}
	model, effort, err := pulseModel(listed["data"])
	if err != nil {
		return nil, err
	}
	started, err := a.Client.Request(ctx, "thread/start", map[string]any{"ephemeral": true, "model": model, "serviceTier": "default"})
	if err != nil {
		return nil, err
	}
	threadID := nestedText(started, "thread", "id")
	if threadID == "" {
		return nil, errors.New("Codex returned no window-pulse thread ID")
	}
	turn, err := a.Client.Request(ctx, "turn/start", map[string]any{"threadId": threadID, "input": []any{map[string]any{"type": "text", "text": "Reply OK."}}, "model": model, "effort": effort, "serviceTier": "default"})
	if err != nil {
		return nil, err
	}
	turnID := nestedText(turn, "turn", "id")
	if turnID == "" {
		return nil, errors.New("Codex returned no window-pulse turn ID")
	}
	for {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-a.Client.Done():
			return nil, a.Client.Err()
		case message := <-a.Client.Notifications():
			if text(message["method"]) != "turn/completed" {
				continue
			}
			params, _ := message["params"].(map[string]any)
			completed, _ := params["turn"].(map[string]any)
			if text(completed["id"]) != turnID {
				continue
			}
			if text(completed["status"]) != "completed" {
				return nil, errors.New("Codex window pulse failed")
			}
			return a.RateLimits(ctx)
		}
	}
}

func pulseModel(raw any) (string, string, error) {
	type candidate struct {
		rank          int
		model, effort string
	}
	var candidates []candidate
	items, _ := raw.([]any)
	for _, rawItem := range items {
		item, _ := rawItem.(map[string]any)
		if item == nil || item["hidden"] == true || text(item["model"]) == "" || !listHas(item["inputModalities"], "text") {
			continue
		}
		var efforts []string
		for _, rawEffort := range list(item["supportedReasoningEfforts"]) {
			if value, ok := rawEffort.(map[string]any); ok && text(value["reasoningEffort"]) != "" {
				efforts = append(efforts, text(value["reasoningEffort"]))
			}
		}
		if len(efforts) == 0 {
			continue
		}
		effort := efforts[0]
		if has(efforts, "minimal") {
			effort = "minimal"
		} else if has(efforts, "low") {
			effort = "low"
		}
		rank := 1
		if contains(text(item["model"]), "mini") {
			rank = 0
		}
		candidates = append(candidates, candidate{rank, text(item["model"]), effort})
	}
	if len(candidates) == 0 {
		return "", "", errors.New("Codex returned no text model for window pulse")
	}
	sort.Slice(candidates, func(i, j int) bool {
		if candidates[i].rank != candidates[j].rank {
			return candidates[i].rank < candidates[j].rank
		}
		return candidates[i].model < candidates[j].model
	})
	return candidates[0].model, candidates[0].effort, nil
}

func text(value any) string {
	result, _ := value.(string)
	return result
}
func integer(value any) int64 {
	if v, ok := value.(float64); ok {
		return int64(v)
	}
	return 0
}
func list(value any) []any {
	result, _ := value.([]any)
	return result
}
func listHas(value any, wanted string) bool {
	for _, item := range list(value) {
		if text(item) == wanted {
			return true
		}
	}
	return false
}
func has(values []string, wanted string) bool {
	for _, item := range values {
		if item == wanted {
			return true
		}
	}
	return false
}
func nestedText(value map[string]any, outer, inner string) string {
	nested, _ := value[outer].(map[string]any)
	if nested == nil {
		return ""
	}
	return text(nested[inner])
}
